src/llama.cpp

```diff
@@ -3687,10 +3687,8 @@ static bool llama_kv_cache_seq_rm(
                 const llama_kv_cell & cell = cache.cells[tail_id];
                 // partial intersection is invalid
                 if ((0 < p0 && p0 <= cell.pos) || (0 < p1 && p1 <= cell.pos)) {
-                    return false;
-                }
-                if (p0 <= cell.pos && p1 < cell.pos) {
                     tail_id = -1;
+                    return false;
                 }
```

Comment:
This does not fix the problem. Building `llama-server` from this branch, I still get

```
llama_kv_cache_find_slot: wrong tail for seq_id 1, (0 instead of -1)
/.../src/llama.cpp:3507: GGML_ASSERT(cell.has_seq_id(seq_id)) failed
```
When using the first request from <https://github.com/ggerganov/llama.cpp/issues/9170> with `curl` and then using the server Web UI for something unrelated. I'll investigate the root cause, but it **does** seem like the tail is not invalidated properly somewhere (which by the way also *has* to be done in sync with clearing the actual tail cell too, which is why the tail was previously invalidated here only when ***not*** returning **and** when the cell (which was the tail) would be cleared with the range conditions below).

Strangely enough, there is no problem when I try using the server UI directly.

Hmm, re-reading my parenthesis above, the problem in the code could actually simply be

```diff
diff --git a/src/llama.cpp b/src/llama.cpp
index 2274296b..47c9a4cb 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -3689,7 +3689,7 @@ static bool llama_kv_cache_seq_rm(
                 if ((0 < p0 && p0 <= cell.pos) || (0 < p1 && p1 <= cell.pos)) {
                     return false;
                 }
-                if (p0 <= cell.pos && p1 < cell.pos) {
+                if (p0 <= cell.pos && cell.pos < p1) {
                     tail_id = -1;
                 }
             }
```
(when diffed from `master`)

Re-running the requests from <https://github.com/ggerganov/llama.cpp/issues/9170> no longer seems to fail with this.

Reply:
Thanks the feedback! I haven't tried this through `curl`, and don't know why there are different behavior between `curl` and web ui, maybe because of the different times of response? I have seen the times of response maybe inconsistent for the same request(same parameters).